src/optimizer/analyze.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>
+#include <cstring>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <random>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "optimizer/analyze.h"
+
+namespace terrier {
+
+template <typename T>
+terrier::optimizer::ColumnStats CalculateColumnStats(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
+                                                     catalog::col_oid_t columnID, storage::SqlTable *sql_table,
+                                                     terrier::transaction::TransactionContext *scan_txn,
+                                                     unsigned mostFrequentNumber) {
+  // Stats collecting temporary data-structures
+  std::unordered_map<T, unsigned> counting_map;
+  unsigned num_null_values = 0;
+  unsigned num_rows = 0;
+  unsigned top_k = mostFrequentNumber;
+  std::vector<double> histogram_bounds;
+  histogram_bounds.push_back(DBL_MAX);
+  histogram_bounds.push_back(DBL_MIN);
+
+  std::vector<catalog::col_oid_t> current_col_oid_vector{columnID};
+  storage::ProjectionMap projection_list_indices = sql_table->ProjectionMapForOids(current_col_oid_vector);
+  auto initializer =
+      sql_table->InitializerForProjectedColumns(current_col_oid_vector, common::Constants::K_DEFAULT_VECTOR_SIZE);
+  auto *buffer = common::AllocationUtil::AllocateAligned(initializer.ProjectedColumnsSize());
+  storage::ProjectedColumns *projected_column = initializer.Initialize(buffer);
+
+  // scan the results
+  auto iterator_table = sql_table->begin();
+  while (iterator_table != sql_table->end()) {
+    sql_table->Scan(scan_txn, &iterator_table, projected_column);
+    common::RawBitmap *null_bit_map = projected_column->ColumnNullBitmap(projection_list_indices[columnID]);
+    T *col_pointer = reinterpret_cast<T *>(projected_column->ColumnStart(projection_list_indices[columnID]));
+    for (uint32_t i = 0; i < projected_column->NumTuples(); i++) {
+      if (null_bit_map->Test(i)) {
+        if (counting_map.find(col_pointer[i]) != counting_map.end())
+          counting_map[col_pointer[i]]++;
+        else
+          counting_map[col_pointer[i]] = 1;
```

Comment:
Are you storing every unique value in this map? What if I have 1 billion unique values?